### ISSUE TITLE
DeDRM fails to decrypt certain free eBooks via Kindle for Android.

### ISSUE THREAD

**Issue by WildSnail913** [noDRM/DeDRM_tools#351]
### Question / bug report

This is the book: https://www.amazon.com/dp/B09GKVC9JS

DeDRM works when decrypting this book with a Kindle eReader download and its Serial Number.

DeDRM fails when decrypting this book with a Kindle for Android download and backup.ab file. 

I am able to successfully DeDRM other books from my Kindle for Android app, just not this one. Given the much higher image quality from the Android download I'd prefer to backup that one over the eReader file.

### Which version of Calibre are you running?

5.44.0

### Which version of the DeDRM plugin are you running?

v10.0.3

### If applicable, which version of the Kindle software are you running?

Kindle for Android 8.51.1.0

### Log output

```log
DeDRM v10.0.3: Trying to decrypt B09GKVC9JS_EBOK.prc
Using Library AlfCrypto DLL/DYLIB/SO
MobiDeDrm v1.1.
Copyright © 2008-2020 The Dark Reverser, Apprentice Harper et al.
Decrypting Kindle Format 8 ebook: Remember the Hand (Fordham Series in Medieval Studies)
Got DSN key from database default_key
Found 12 keys to try after 0.1 seconds
Crypto Type is: 2
DeDRM v10.0.3: Failed to decrypt with error: No key found in 12 PIDs tried.
DeDRM v10.0.3: Looking for new default Kindle Key after 0.1 seconds
searching for kinfoFiles in C:\Users\WildSnail\AppData\Local
No K4PC kindle.info/kinf/kinf2011 files have been found.
DeDRM v10.0.3: Ultimately failed to decrypt after 0.1 seconds. Read the FAQs at noDRM's repository: 
```

**Comment by ElleKayEm** [comment]
What size was your backup file?

**Comment by WildSnail913** [comment]
> What size was your backup file?

50.3 MB. It works for other books off Android, so I assumed the backup.ab key was not the problem. If you have an Android setup the book is free so you can give it a try if you like.

**Comment by ElleKayEm** [comment]
I haven't tried the Kindle for Android method yet myself.  Yeah, if the key works for other books downloaded with this setup, then that's not the issue.  Just noticed the prc extension on that file, and I wonder if you might have luck with older versions of DeDRM and calibre.  Some of the less frequently seen formats haven't gotten much testing on the later DeDRM.

DeDRM 6.8.1: https://github.com/apprenticeharper/DeDRM_tools/releases/tag/v6.8.1

calibre 4.23:  https://download.calibre-ebook.com/4.23.0/

**Comment by WildSnail913** [comment]
> might have luck with older versions of DeDRM and calibre

Tried it with Calibre 4.23 Portable and DeDRM 6.8.0 without success. The ebook in the first post still failed to decrypt. My other books off Android still decrypted fine.

Also, I noticed a couple other free books fail to decrypt off Kindle for Android as well:
https://www.amazon.com/dp/B074SGMCL7
https://www.amazon.com/dp/B0C7R4NQ16

Commonality between these is that theyre all image heavy. Maybe its something to do with that?

I tried a random free novel, without a lot of images, and it decrypted fine off Kindle for Android.

**Comment by ableeker** [comment]
The backup.ab unfortunately doesn't work no longer, because you can't create backups anymore. When you do have a working backup.ab, DeDRM will extract data from it, and in the end generate keys that are certain combinations of the extracted data. All combinations found are used to try to decrypt the ebook, and if one of them succeeds, we've have found the key. This works for the ebooks I have, but it seems not for all of them out there. I've added some more combinations in the enclosed DeDRM plugin. Could you try this version to see if this one is working?

[DeDRM_plugin.zip](https://github.com/noDRM/DeDRM_tools/files/11867664/DeDRM_plugin.zip)

**Comment by j-howell** [comment]
> I am able to successfully DeDRM other books from my Kindle for Android app, just not this one.

DRM removal apparently working randomly on some books and not on others is usually a sign that the plugin is not configured with the correct key and that the ones that worked did not have DRM in the first place. Examples of some books that succeeded would let us know whether or not that is the case.

> If applicable, which version of the Kindle software are you running?
> Kindle for Android 8.51.1.0

The last version of Kindle for Android that will produce a usable backup is 4.17.0.146. Later versions of the Kindle app are coded to prevent backup.

If you want to use version 8.51.1.0 you first need to install an older version, obtain the key, and then update the app in place so that the original key is retained by the new app version. Instructions for doing that can be found in [this post](https://www.mobileread.com/forums/showthread.php?p=4305731#post4305731) on MobileRead.

**Comment by j-howell** [comment]
> When you do have a working backup.ab, DeDRM will extract data from it, and in the end generate keys that are certain combinations of the extracted data. All combinations found are used to try to decrypt the ebook, and if one of them succeeds, we've have found the key. This works for the ebooks I have, but it seems not for all of them out there. I've added some more combinations in the enclosed DeDRM plugin.

The key needed for the Kindle for Android app is always a serial number followed by an account token. There is no point in trying combinations other than those already generated by the plugin.

**Comment by cgaspar** [comment]
I can confirm I see the same issue using Android Kindle 4.16.0.75. Extracting the keys via backup works, other books decrypt fine, one does not. Same result with the plugin provided in https://github.com/noDRM/DeDRM_tools/issues/351#issuecomment-1607211244. 

However, I CAN decrypt https://www.amazon.com/dp/B09HTBQ93K/, unlike the original reporter.

**Comment by ElleKayEm** [comment]
> > might have luck with older versions of DeDRM and calibre
> 
> Tried it with Calibre 4.23 Portable and DeDRM 6.8.1 without success. The ebook in the first post still failed to decrypt. My other books off Android still decrypted fine.
> 

With large files like that, you want 64-bit calibre.

**Comment by j-howell** [comment]
> I can confirm I see the same issue using Android Kindle 4.16.0.75. Extracting the keys via backup works, other books decrypt fine, one does not. Same result with the plugin provided in [#351 (comment)](https://github.com/noDRM/DeDRM_tools/issues/351#issuecomment-1607211244).
> 
> However, I CAN decrypt https://www.amazon.com/dp/B09HTBQ93K/, unlike the original reporter.

That book does have DRM so you must have the proper key installed in the plugin.

Most likely the book that does not work is in a format that DeDRM doesn’t handle properly. I believe that there are problems with Topaz books in the current release of the plugin.  If you share which book fails then we can determine if that is the case.

**Comment by cgaspar** [comment]
Failing book: https://www.amazon.com//dp/B0C4R4V6KN

**Comment by ElleKayEm** [comment]
Purchase or borrow?  And which versions of DeDRM and calibre?

**Comment by cgaspar** [comment]
Looks like this one is kindle unlimited. DeDRM 10.0.2 (from comment above), Calibre 6.21. So far I have exactly one book that won't decrypt - other purchased and unlimited books have all worked. So $something has changed on the Amazon side.

**Comment by ElleKayEm** [comment]
So borrowed then?  What's the file extension?  Use DeDRM 10.0.3 with calibre 6.

**Comment by cgaspar** [comment]
10.0.3 fails (what I had originally - not sure why the version posted in the link above shows as 10.0.2...)

**Comment by ElleKayEm** [comment]
You might try the older versions I mention above but that DeDRM may block borrowed books.

FYI, read it in the Kindle app or buy it if you want the author to get paid.

**Comment by WildSnail913** [comment]
> If you want to use version 8.51.1.0 you first need to install an older version, obtain the key, and then update the app in place so that the original key is retained by the new app version.

I did get the key from the older version however I DID NOT update the app in place. (made a new instance)

I redid this process while updating in place this time and was able to successfully DeDRM both one of the books I had trouble with, https://www.amazon.com/dp/B09HTBQ93K/ , and the book cgaspar was having trouble with https://www.amazon.com//dp/B0C4R4V6KN .  Not sure why the latter worked for me but not csgaspar but there you go. (I used Calibre 5.44 64 bit, NoDRM's DeDRM 10.0.3)

I checked using the drmcheck.py script before importing and both of these files were "encrypted with the current encryption method". Checked a couple of the books I "successfully" decrypted before and they were unecrypted without importing.

Sorry for the big old user error, thanks for the help though. I'll mark this issue as closed.
